### PR TITLE
An employee can't set for himself or other a permission he does not have

### LINKED ISSUE #10216: [BOOM-4689] Permission management error
http://forge.prestashop.com/browse/BOOM-4689

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | As an employee, I'm able to change my own permissions and others permissions even if I don't have these permissions formerly. This defeat the principle of permissions management :/
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/10216
| How to test?  | Play with permissions in Advanced Parameters > Team > Permissions.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11196)

### CODE REVIEW

File: controllers/admin/AdminAccessController.php [hunk 1]
```diff
@@ -140,14 +140,16 @@ public function ajaxProcessUpdateAccess()
         if (_PS_MODE_DEMO_) {
             throw new PrestaShopException($this->trans('This functionality has been disabled.', array(), 'Admin.Notifications.Error'));
         }
-        if ($this->access('edit') != '1') {
+
+        $tab_id = (int) Tools::getValue('id_tab');
+        $permission = Tools::getValue('perm');
+        if ($this->isAllowedToEditPermissions($tab_id, $permission)) {
```
Comment (matks) [inline controllers/admin/AdminAccessController.php]: Looks like a `!` is missing 🤔 ?

File: controllers/admin/AdminAccessController.php [hunk 2]
```diff
@@ -223,4 +226,34 @@ private function getChildrenTab(array &$tabs, $id_parent = 0)
 
         return $children;
     }
+
+    /**
+     * @param $tab_id
+     * @param $permission
```
Comment (matks) [inline controllers/admin/AdminAccessController.php]: We do not know what `$permission` is ?

**Review by matks — DISMISSED** [review]

File: controllers/admin/AdminAccessController.php [hunk 3]
```diff
@@ -223,4 +226,34 @@ private function getChildrenTab(array &$tabs, $id_parent = 0)
...
+        $currentProfile = (int) ContextCore::getContext()->employee->id_profile;
+        $slug = Access::findSlugByIdTab($tab_id);
+
+        foreach ((array) Access::getAuthorizationFromLegacy($permission) as $legacyAuthorization) {
+            if (!Access::isGranted($slug . $legacyAuthorization, $currentProfile)) {
+                return false;
+            }
+        }
+
+        if (!$this->access('edit')) {
```
Comment (PierreRambaud) [inline controllers/admin/AdminAccessController.php]: `return  $this->access('edit');` :)

**Comment by mickaelandrieu** [comment]
ping @matks @PierreRambaud: comments addressed!

File: controllers/admin/AdminAccessController.php [hunk 4]
```diff
@@ -223,4 +226,30 @@ private function getChildrenTab(array &$tabs, $id_parent = 0)
...
+        }
+
+        $currentProfile = (int) ContextCore::getContext()->employee->id_profile;
+        $slug = Access::findSlugByIdTab($tab_id);
+
+        foreach ((array) Access::getAuthorizationFromLegacy($permission) as $legacyAuthorization) {
+            if (!Access::isGranted($slug . $legacyAuthorization, $currentProfile)) {
+                return false;
+            }
+        }
```
Comment (Quetzacoalt91) [inline controllers/admin/AdminAccessController.php:263]: Could you please add some comments on what you're checking here?

Comment (mickaelandrieu) [reply controllers/admin/AdminAccessController.php:263]: I don't know what this function does under the hood.

Comment (mickaelandrieu) [reply controllers/admin/AdminAccessController.php:263]: to be more precise, I don't know for sure what does do Access::isGranted() so I've done some assumptions about it and it works as expected 😇

Comment (Quetzacoalt91) [reply controllers/admin/AdminAccessController.php:263]: I added the comments by myself. We could not approve stuff we don't understand.

**Review by Quetzacoalt91 — DISMISSED** [review]

**Comment by marionf** [comment]
@mickaelandrieu 

This issue seems already solved on 1.7.5.x
I can't give to myself permissions that I don't have

Checkboxes view, add, edit, delete are disabled

capture d ecran_678

Checkboxes not checked are disabled also

capture d ecran_679

This fix is really necessary ?

**Comment by marionf** [comment]
@mickaelandrieu 

I can disable my own permissions but not re-enable it after

https://drive.google.com/open?id=1CPZ7S4zPTSTpoic1nFU7tlYIlawWvt_h

**Comment by mickaelandrieu** [comment]
> I can disable my own permissions but not re-enable it after

I don't think we can change this behavior before changing dramatically how permissions system works, sadly :/

This needs to be discussed with @Florinette and implemented in 1.7.6: this is not a bug but the intended behavior. 
There is no way to distinguish the authorizations you don't have (and so on you're not allowed to have) from the ones you used to have before and want to restore.

I guess this can be solved by a new action named "Restore my authorizations from my role" but this is not something implemented (even partially) right now.

In fact, I don't understand why a user can update its owns permissions at all: you can't add new permissions and you have no concrete reason to remove permissions too.

**Comment by marionf** [comment]
Currently, I think this may be problematic because if a user inadvertenly removes a permission, he can't put it back

**Comment by marionf** [comment]
@mickaelandrieu suggest to put the permission page in read only mode for your own profile

@colinegin wdyt ?

**Comment by Florinette** [comment]
> I don't think we can change this behavior before changing dramatically how permissions system works, sadly :/
> 
> This needs to be discussed with @Florinette and implemented in 1.7.6: this is not a bug but the intended behavior.
> There is no way to distinguish the authorizations you don't have (and so on you're not allowed to have) from the ones you used to have before and want to restore.
> 
> I guess this can be solved by a new action named "Restore my authorizations from my role" but this is not something implemented (even partially) right now.
> 
> In fact, I don't understand why a user can update its owns permissions at all: you can't add new permissions and you have no concrete reason to remove permissions too.

I think you probably should not be able to remove your own permissions... that avoids the problem that you can't add them back.

**Comment by marionf** [comment]
Can we add something more clear than "Unable to update settings" ?

capture d ecran_686

Like on the top of the page a message "You can't modify your own permissions" ?

I think we should also move this one to develop branch for 1.7.6 version because specifications are not very clear, it's not a regression and we are a few days before the 1.7.5 release
Maybe we should take time to think more about the wanted behavior

File: controllers/admin/AdminAccessController.php [hunk 5]
```diff
@@ -223,4 +226,45 @@ private function getChildrenTab(array &$tabs, $id_parent = 0)
...
+        if ($tab_id === -1) {
+            return false;
+        }
+
+        $currentProfile = (int) ContextCore::getContext()->employee->id_profile;
+
+        /**
+         * The current employee can't change its own permissions
+         */
+        if ($profile_id === $currentProfile) {
```
Comment (jolelievre) [inline controllers/admin/AdminAccessController.php:250]: so an admin or super admin user can't even edit his own permissions?

**Review by jolelievre — CHANGES_REQUESTED** [review]

File: controllers/admin/AdminAccessController.php (hunk 5 above)
Comment (mickaelandrieu) [reply controllers/admin/AdminAccessController.php:250]: yes, it's the idea! See @Florinette and @marionf comments

**Comment by kpodemski** [comment]
It's security issue and it should be part of 1.7.5, PrestaShop Ready is affected

**Review by PierreRambaud — DISMISSED** [review]

**Comment by marionf** [comment]
@mickaelandrieu as this one isn't ready for 1.7.5, can you rebase on develop ?

I saw with @TristanLDD & @colinegin the error message should be "You do not have permission to edit this" instead of "Unable to update settings" 

& if we can add an information message "You can't modify your own permissions" on the top of the page when you are on your own profile 

capture d ecran_688

@LouiseBonnard Are you ok with this new string ?

**Comment by PierreRambaud** [comment]
Let's fix the wording and move the target to develop :)

**Comment by LouiseBonnard** [comment]
@marionf, it's alright for me, thanks for the comment! But wouldn't it be nice to help the user and let him know how to modify it?

**Comment by mickaelandrieu** [comment]
If the objective is to fix it for the 1.7.6.x, I don't see the value implementing a new way of setting a new error message in legacy when this page is probably part of the migration and we are already able to set the right error message in the migrated page...

Ping @matks is this page part of the 1.7.6.x roadmap?

**Comment by mickaelandrieu** [comment]
> It's security issue and it should be part of 1.7.5, PrestaShop Ready is affected

It's not a security issue, it's a terrible UX issue. Fix is already done in PrestaShop Ready (not the same fix).

**Comment by LouiseBonnard** [comment]
Hi there, in order to get a more helpful information, I suggest the following message: _Sorry, you can't modify your own permissions, you need to ask the SuperAdmin_. What do you think, is it precise enough?
@colinegin, @marionf, @TristanLDD

**Comment by colinegin** [comment]
Wording is ok with the product team !

**Comment by mickaelandrieu** [comment]
> Ping @matks is this page part of the 1.7.6.x roadmap?

Really important :)

**Comment by matks** [comment]
@mickaelandrieu Yes, it's in the 5 next pages to be migrated.

**Comment by mickaelandrieu** [comment]
Thanks @matks, so I don't think it's worth to spend time to fix the bug (the error message in legacy, while correctly set up, is never displayed).

Can you link this pull request to the issue of migration so the developer can use the "right" error message in the new page? 👼 

Thanks mate!

**Comment by matks** [comment]
Here we go 😉 https://github.com/PrestaShop/PrestaShop/issues/10504

**Comment by Quetzacoalt91** [comment]
Following the last messages, this PR is now closed as the page will be shortly migrated.

**Pull request closed without merging**